File: iosApp/iosApp/Security/ReownWalletManager.swift
//
// ReownWalletManager.swift
// iosApp
//
// Manager for orchestrating Reown AppKit operations.
// Provides high-level methods for wallet connection and signing.
// Based on AppKitLab example from reown-swift
//

import Foundation
import Combine
import ReownAppKit
import idos_sdk
import os.log

/// Manager for Reown AppKit operations
class ReownWalletManager {
    private let logger = Logger(subsystem: "org.idos.app", category: "ReownWalletManager")
    private let projectId: String
    private var isInitialized = false
    private var cancellables = Set<AnyCancellable>()
    private var pendingPersonalSignContinuation: CheckedContinuation<String, Error>?

    init(projectId: String) {
        self.projectId = projectId
    }

    /// Initialize Reown SDK (AppKit).
    /// Must be called before any other operations.
    func initialize() {
        guard !isInitialized else {
            logger.debug("ReownWalletManager already initialized")
            return
        }

        let metadata = AppMetadata(
            name: "idOS SDK",
            description: "",
            url: "https://idos.network",
            icons: ["https://idos.network/icon.png"],
            redirect: try! AppMetadata.Redirect(native: "idos-app://", universal: nil)
        )

        // Configure Networking first (required by AppKit)
        // Use app group identifier (must start with "group.")
        Networking.configure(
            groupIdentifier: "group.org.idos.app",
            projectId: projectId,
            socketFactory: DefaultSocketFactory()
        )

        // Configure AppKit with proper parameters
        AppKit.configure(
            projectId: projectId,
            metadata: metadata,
            crypto: DefaultCryptoProvider(),
            authRequestParams: nil  // Set to .stub() if SIWE authentication is needed
        ) { error in
            self.logger.error("AppKit configuration error: \(error)")
        }

        // Subscribe to session responses
        AppKit.instance.sessionResponsePublisher
            .sink { [weak self] response in
                self?.handleSessionResponse(response)
            }
            .store(in: &cancellables)

        // Subscribe to socket connection status
        AppKit.instance.socketConnectionStatusPublisher
            .receive(on: DispatchQueue.main)
            .sink { [weak self] status in
                self?.logger.debug("Socket connection status: \(String(describing: status))")
            }
            .store(in: &cancellables)

        // Set logging level
        #if DEBUG
        AppKit.instance.logger.setLogging(level: .debug)
        #else
        AppKit.instance.logger.setLogging(level: .error)
        #endif

        // Initialize delegate for additional event handling
        ReownDelegate.shared.initialize()

        isInitialized = true
        logger.debug("ReownWalletManager initialized")
    }

    /// Get the currently connected wallet address
    func getConnectedAddress() -> String? {
        return AppKit.instance.getAddress()
    }

    /// Send a personal_sign request to the connected wallet
    ///
    /// - Parameter message: Message to sign (raw bytes)
    /// - Returns: Signature as hex string
    func personalSign(message: Data) async throws -> String {
        guard let address = getConnectedAddress() else {
            throw WalletError.notConnected
        }

        let messageString = String(data: message, encoding: .utf8) ?? message.hexString

        // Create a continuation to wait for the response
        return try await withCheckedThrowingContinuation { continuation in
            // Store continuation to be resumed when response arrives
            pendingPersonalSignContinuation = continuation

            // Send the request (returns Void)
            Task {
                do {
                    try await AppKit.instance.request(
                        .personal_sign(address: address, message: messageString)
                    )
                } catch {
                    pendingPersonalSignContinuation = nil
                    continuation.resume(throwing: error)
                }
            }
        }
    }

    /// Send an eth_signTypedData_v4 request to the connected wallet
    ///
    /// - Parameter typedData: EIP-712 typed data to sign
    /// - Returns: Signature as hex string
    func signTypedData(_ typedData: TypedData) async throws -> String {
        guard let address = getConnectedAddress() else {
            throw WalletError.notConnected
        }

        // Use TypedData's built-in JSON serialization
        let typedDataJsonString = typedData.toJsonString()

        // Note: W3MJSONRPC doesn't have eth_signTypedData_v4
        // We'll need to use personal_sign as a fallback or wait for SDK support
        // For now, throw an error indicating this isn't supported yet
        logger.error("eth_signTypedData_v4 not yet supported in iOS AppKit")
        throw WalletError.requestFailed("eth_signTypedData_v4 not supported - SDK limitation")
    }

    /// Disconnect from the current wallet session
    func disconnect() async throws {
        // TODO: Check AppKit disconnect API - may need session topic
        // try await AppKit.instance.disconnect(topic: selectedTopic)
        logger.debug("Disconnect not yet implemented")
    }

    // MARK: - Private Methods

    private func handleSessionResponse(_ response: W3MResponse) {
        switch response.result {
        case .response(let value):
            logger.debug("Session response received: \(value.stringRepresentation)")

            // Resume pending personal_sign continuation if exists
            if let continuation = pendingPersonalSignContinuation {
                pendingPersonalSignContinuation = nil
                if let stringValue = value.value as? String {
                    continuation.resume(returning: stringValue)
                } else {
                    continuation.resume(throwing: WalletError.requestFailed("Invalid response type"))
                }
            }

            // Also notify delegate for other listeners
            if let stringValue = value.value as? String {
                ReownDelegate.shared.handleRequestResponse(result: stringValue)
            }
        case .error(let error):
            logger.error("Session error: \(error.message)")

            // Resume pending continuation with error if exists
            if let continuation = pendingPersonalSignContinuation {
                pendingPersonalSignContinuation = nil
                continuation.resume(throwing: WalletError.requestFailed(error.message))
            }

            // Also notify delegate
            ReownDelegate.shared.handleRequestError(WalletError.requestFailed(error.message))
        }
    }
}

// MARK: - WalletError
enum WalletError: LocalizedError {
    case notConnected
    case requestFailed(String)
    case requestExpired
    case sendFailed(String)
    case serializationFailed

    var errorDescription: String? {
        switch self {
        case .notConnected:
            return "Wallet not connected"
        case .requestFailed(let message):
            return "Request failed: \(message)"
        case .requestExpired:
            return "Request expired"
        case .sendFailed(let message):
            return "Failed to send request: \(message)"
        case .serializationFailed:
            return "Failed to serialize data"
        }
    }
}

// MARK: - Data Extension
private extension Data {
    var hexString: String {
        map { String(format: "%02x", $0) }.joined()
    }
}

File: androidApp/src/main/java/org/idos/app/security/external/ReownWalletManager.kt
package org.idos.app.security.external

import android.app.Application
import com.reown.android.Core
import com.reown.android.CoreClient
import com.reown.appkit.client.AppKit
import com.reown.appkit.client.Modal
import com.reown.appkit.client.models.request.Request
import com.reown.appkit.presets.AppKitChainsPresets
import com.reown.appkit.utils.EthUtils
import com.reown.util.bytesToHex
import com.reown.util.randomBytes
import kotlinx.coroutines.CoroutineScope
import kotlinx.coroutines.Dispatchers
import kotlinx.coroutines.SupervisorJob
import kotlinx.coroutines.flow.filter
import kotlinx.coroutines.launch
import kotlinx.serialization.encodeToString
import kotlinx.serialization.json.Json
import org.idos.crypto.eip712.TypedData
import timber.log.Timber
import kotlin.coroutines.resume
import kotlin.coroutines.resumeWithException
import kotlin.coroutines.suspendCoroutine

/**
 * Manager for orchestrating Reown AppKit operations.
 * Provides high-level methods for wallet connection and signing.
 */
class ReownWalletManager(
    private val application: Application,
    private val projectId: String
) {
    private val scope = CoroutineScope(SupervisorJob() + Dispatchers.IO)
    private var isInitialized = false

    /**
     * Initialize Reown SDK (CoreClient and AppKit).
     * Must be called before any other operations.
     */
    fun initialize() {
        if (isInitialized) {
            Timber.d("ReownWalletManager already initialized")
            return
        }

        val appMetaData = Core.Model.AppMetaData(
            name = "idOS SDK",
            description = "",
            url = "https://idos.network",
            icons = listOf("https://idos.network/icon.png"),
            redirect = "idos-app://request",
//            linkMode = true
        )

        CoreClient.initialize(
            application = application,
            projectId = projectId,
            metaData = appMetaData,
        ) {
            Timber.e(it.throwable, "CoreClient initialization failed")
        }

        AppKit.initialize(Modal.Params.Init(core = CoreClient)) { error ->
            Timber.e(error.throwable, "AppKit initialization failed")
        }

        AppKit.setChains(AppKitChainsPresets.ethChains.values.toList())

        // Initialize delegate
        ReownDelegate.initialize()

        isInitialized = true
        Timber.d("ReownWalletManager initialized")
    }

    /**
     * Get the currently connected wallet address.
     */
    fun getConnectedAddress(): String? {
        return AppKit.getAccount()?.address
    }

    /**
     * Send a personal_sign request to the connected wallet.
     *
     * @param message Message to sign (will be hex encoded)
     * @return Signature as hex string
     */
    suspend fun personalSign(message: ByteArray): String {
        val address = getConnectedAddress()
            ?: throw IllegalStateException("Wallet not connected")

        val messageHex = "0x" + message.bytesToHex()

        val requestParams = Request(
            method = "personal_sign",
            params = """["$messageHex", "$address"]"""
        )

        return sendWalletRequest(requestParams, "personal_sign")
    }

    /**
     * Send an eth_signTypedData_v4 request to the connected wallet.
     *
     * @param typedData EIP-712 typed data to sign
     * @return Signature as hex string
     */
    suspend fun signTypedData(typedData: TypedData): String {
        val address = getConnectedAddress()
            ?: throw IllegalStateException("Wallet not connected")

        // Serialize TypedData to JSON string
        val escapedTypedData = typedData.toJsonString()

        val requestParams = Request(
            method = "eth_signTypedData_v4",
            params = """["$address", $escapedTypedData]"""
        )

        return sendWalletRequest(requestParams, "eth_signTypedData_v4")
    }

    /**
     * Disconnect from the current wallet session.
     */
    fun disconnect() {
        try {
            AppKit.disconnect(
                onSuccess = {
                    Timber.d("Disconnected from wallet")
                },
                onError = { error ->
                    Timber.e(error, "Failed to disconnect from wallet")
                }
            )
        } catch (e: Exception) {
            Timber.e(e, "Failed to disconnect")
        }
    }

    /**
     * Send a request to the wallet and await response.
     * Common handler for all wallet signing operations.
     *
     * @param requestParams The request to send
     * @param requestName Human-readable name for logging (e.g., "personal_sign")
     * @return Result string from the wallet
     */
    private suspend fun sendWalletRequest(
        requestParams: Request,
        requestName: String
    ): String = suspendCoroutine { continuation ->
        Timber.d("Sending $requestName request")

        var isResumed = false

        // Collect events in coroutine scope
        scope.launch {
            ReownDelegate.appKitEvents
                .filter { it is Modal.Model.SessionRequestResponse || it is Modal.Model.ExpiredRequest }
                .collect { event ->
                    if (isResumed) return@collect

                    when (event) {
                        is Modal.Model.SessionRequestResponse -> {
                            when (val result = event.result) {
                                is Modal.Model.JsonRpcResponse.JsonRpcResult -> {
                                    Timber.d("$requestName request successful")
                                    isResumed = true
                                    continuation.resume(result.result ?: "")
                                }
                                is Modal.Model.JsonRpcResponse.JsonRpcError -> {
                                    Timber.e("$requestName request failed: ${result.message}")
                                    isResumed = true
                                    continuation.resumeWithException(
                                        Exception("$requestName request failed: ${result.message}")
                                    )
                                }
                            }
                        }
                        is Modal.Model.ExpiredRequest -> {
                            Timber.e("$requestName request expired")
                            isResumed = true
                            continuation.resumeWithException(
                                Exception("$requestName request expired")
                            )
                        }
                        else -> {}
                    }
                }
        }

        // Send the request
        AppKit.request(
            request = requestParams,
            onSuccess = { _ ->
                Timber.d("Request sent successfully, awaiting response from delegate")
            },
            onError = { error: Throwable ->
                Timber.e(error, "Failed to send $requestName request")
                if (!isResumed) {
                    isResumed = true
                    continuation.resumeWithException(
                        Exception("Failed to send $requestName request: ${error.message}")
                    )
                }
            }
        )
    }
}
